File: Benchmark.java
import java.util.ArrayList;
import java.util.List;

public class Benchmark {
    public static void main(String[] args) {
        int N = 1000000;
        List<String> titles = new ArrayList<>(N);
        for (int i = 0; i < N; i++) {
            titles.add("my beautiful title " + i);
        }
        String query = "Beautiful";
        List<String> words = new ArrayList<>();
        words.add("beautiful");

        // Warmup
        for (int i = 0; i < 10; i++) {
            benchmarkSlow(titles, query, words);
            benchmarkFast(titles, query, words);
        }

        long t0 = System.nanoTime();
        benchmarkSlow(titles, query, words);
        long t1 = System.nanoTime();

        long t2 = System.nanoTime();
        benchmarkFast(titles, query, words);
        long t3 = System.nanoTime();

        System.out.println("Baseline: " + (t1 - t0) / 1000000.0 + " ms");
        System.out.println("Optimized: " + (t3 - t2) / 1000000.0 + " ms");
    }

    public static void benchmarkSlow(List<String> keys, String query, List<String> words) {
        int count = 0;
        for (String title : keys) {
            for (String word : words) {
                if (title.contains(word)) {
                    if (title.contains(query.toLowerCase())) {
                        count++;
                    } else {
                        count++;
                    }
                    break;
                }
            }
        }
    }

    public static void benchmarkFast(List<String> keys, String query, List<String> words) {
        int count = 0;
        String queryLowercase = query.toLowerCase();
        for (String title : keys) {
            for (String word : words) {
                if (title.contains(word)) {
                    if (title.contains(queryLowercase)) {
                        count++;
                    } else {
                        count++;
                    }
                    break;
                }
            }
        }
    }
}

File: benchmark.kt
import kotlin.system.measureNanoTime

fun main() {
    val titleMapKeys = (1..100000).map { "Title $it" }

    val query = "Title 5000"
    val words = query.split(" ").map { it.trim().lowercase() }.filter { it.length > 1 }

    // warm up
    for(i in 1..10) {
        benchmarkSlow(titleMapKeys, query, words)
        benchmarkFast(titleMapKeys, query, words)
    }

    val slowTime = measureNanoTime {
        benchmarkSlow(titleMapKeys, query, words)
    }

    val fastTime = measureNanoTime {
        benchmarkFast(titleMapKeys, query, words)
    }

    println("Baseline: ${slowTime / 1_000_000.0} ms")
    println("Optimized: ${fastTime / 1_000_000.0} ms")
}

fun benchmarkSlow(keys: List<String>, query: String, words: List<String>) {
    var count = 0
    keys.forEach { title ->
        for (word in words) {
            if (title.contains(word)) {
                if (title.contains(query.lowercase())) {
                    count++
                } else {
                    count++
                }
                break
            }
        }
    }
}

fun benchmarkFast(keys: List<String>, query: String, words: List<String>) {
    var count = 0
    val queryLowercase = query.lowercase()
    keys.forEach { title ->
        for (word in words) {
            if (title.contains(word)) {
                if (title.contains(queryLowercase)) {
                    count++
                } else {
                    count++
                }
                break
            }
        }
    }
}
